compiler/rustc_lint/src/gpukernel_abi.rs

```diff
@@ -0,0 +1,177 @@
+use std::iter;
+
+use rustc_abi::ExternAbi;
+use rustc_hir::attrs::AttributeKind;
+use rustc_hir::{self as hir, find_attr};
+use rustc_middle::ty;
+use rustc_session::{declare_lint, declare_lint_pass};
+use rustc_span::Span;
+use rustc_span::def_id::LocalDefId;
+
+use crate::lints::{ImproperGpuKernelArg, MissingGpuKernelExportName};
+use crate::{LateContext, LateLintPass, LintContext};
+
+declare_lint! {
+    /// The `improper_gpu_kernel_arg` lint detects incorrect use of types in `gpu-kernel`
+    /// arguments.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,ignore (fails on non-GPU targets)
+    /// #[unsafe(no_mangle)]
+    /// extern "gpu-kernel" fn kernel(_: [i32; 10]) {}
+    /// ```
+    ///
+    /// This will produce:
+    ///
+    /// ```text
+    /// warning: passing type `[i32; 10]` to a function with "gpu-kernel" ABI may have unexpected behavior
+    ///  --> t.rs:2:34
+    ///   |
+    /// 2 | extern "gpu-kernel" fn kernel(_: [i32; 10]) {}
+    ///   |                                  ^^^^^^^^^
+    ///   |
+    ///   = help: use primitive types and raw pointers to get reliable behavior
+    ///   = note: `#[warn(improper_gpu_kernel_arg)]` on by default
+    /// ```
+    ///
+    /// ### Explanation
+    ///
+    /// The compiler has several checks to verify that types used as arguments in `gpu-kernel`
+    /// functions follow certain rules to ensure proper compatibility with the foreign interfaces.
+    /// This lint is issued when it detects a probable mistake in a signature.
+    IMPROPER_GPU_KERNEL_ARG,
+    Warn,
+    "simple arguments of gpu-kernel functions"
```

Comment:
This line should capture the _reason_ for the lint, not what it is checking for, so something like
```suggestion
    "GPU kernel entry points have a limited calling convention"
```